gtsam/sfm/TransferFactor.h

```diff
@@ -30,89 +84,133 @@ namespace gtsam {
-      const std::vector<std::tuple<Point2, Point2, Point2>>& triplets,
-      const SharedNoiseModel& model = nullptr)
-      : NoiseModelFactorN<F, F>(model, key1, key2),
-        key1_(key1),
-        key2_(key2),
+  TransferFactor(EdgeKey edge1, EdgeKey edge2,
+                 const std::vector<Triplet>& triplets,
+                 const SharedNoiseModel& model = nullptr)
+      : Base(model, edge1, edge2),
+        TransferEdges<F>(edge1, edge2),
         triplets_(triplets) {}
 
-  // Create Matrix3 objects based on EdgeKey configurations
-  std::pair<Matrix3, Matrix3> getMatrices(const F& F1, const F& F2) const {
-    // Fill Fca and Fcb based on EdgeKey configurations
-    if (key1_.i() == key2_.i()) {
-      return {F1.matrix(), F2.matrix()};
-    } else if (key1_.i() == key2_.j()) {
-      return {F1.matrix(), F2.matrix().transpose()};
-    } else if (key1_.j() == key2_.i()) {
-      return {F1.matrix().transpose(), F2.matrix()};
-    } else if (key1_.j() == key2_.j()) {
-      return {F1.matrix().transpose(), F2.matrix().transpose()};
-    } else {
-      throw std::runtime_error(
-          "TransferFactor: invalid EdgeKey configuration.");
-    }
-  }
-
-  /// vector of errors returns 2*N vector
+  /// Vector of errors returns 2*N vector.
   Vector evaluateError(const F& F1, const F& F2,
                        OptionalMatrixType H1 = nullptr,
                        OptionalMatrixType H2 = nullptr) const override {
-    std::function<Vector(const F&, const F&)> transfer = [&](const F& F1,
-                                                             const F& F2) {
+    std::function<Vector(const F&, const F&)> errorFunction = [&](const F& f1,
+                                                                  const F& f2) {
       Vector errors(2 * triplets_.size());
       size_t idx = 0;
-      auto [Fca, Fcb] = getMatrices(F1, F2);
-      for (const auto& tuple : triplets_) {
-        const auto& [pa, pb, pc] = tuple;
-        Point2 transferredPoint = EpipolarTransfer(Fca, pa, Fcb, pb);
-        Vector2 error = transferredPoint - pc;
-        errors.segment<2>(idx) = error;
+      auto [Fca, Fcb] = this->getMatrices(f1, f2);
+      for (const auto& [pa, pb, pc] : triplets_) {
+        errors.segment<2>(idx) = EpipolarTransfer(Fca, pa, Fcb, pb) - pc;
         idx += 2;
       }
       return errors;
     };
-    if (H1) *H1 = numericalDerivative21<Vector, F, F>(transfer, F1, F2);
-    if (H2) *H2 = numericalDerivative22<Vector, F, F>(transfer, F1, F2);
-    return transfer(F1, F2);
+
+    if (H1) *H1 = numericalDerivative21(errorFunction, F1, F2);
+    if (H2) *H2 = numericalDerivative22(errorFunction, F1, F2);
+    return errorFunction(F1, F2);
   }
 };
 
+/**
+ * @class EssentialTransferFactor
+ * @brief Transfers points between views using essential matrices, optimizes for
+ * calibrations of the views, as well. Note that the EssentialMatrixFactor4 does
+ * something similar but without transfer.
```

Comment:
I think optimization of the calibrations should be reserved for the transfer factors that leverage the fundamental matrix, and this factor should only optimize values involved with the essential matrix, i.e., the relative orientation and direction.

Reply:
You can use the TransferFactor above, templated on `EssentialMatrix`, to optimize for R,t only. This factor does E and Ks simultaneously. It works very well as the data bears out.

Reply:
I'll think of better name